File: contest/250304-sduwh-newstar/4.cpp
#include <bits/stdc++.h>
#define endl '\n'
#define rep(i,x,y) for(int i=x;i<=y;i++)
using namespace std;
typedef long long ll;
constexpr int maxn = 2e5+10;
constexpr ll mod = 998244353;
void solve() {
    int n,m; cin >> n >> m;
    vector<int> c(n+5);
    rep(i,1,n) cin >> c[i];
    vector<vector<int>> a(n+5);
    rep(i,1,m) {
        int x, y; cin >> x >> y;
        a[x].push_back(y);
        a[y].push_back(x);
    }
    vector<int> nd;
    vector<int> cn(n+5, -1);
    function<void(int)> dfs = [&](int u) {
        // cout << u << endl;
        nd.push_back(u);
        for (auto v : a[u]) {
            if(cn[v] == -1) {
                cn[v] = 1 - cn[u];
                dfs(v);
            } else if (cn[u] == cn[v]) {
                cout << "Impossible" << endl;
                exit(0);
            }
        }
    };
    int ans = 0;
    rep(i,1,n) {
        if(cn[i] == -1) {
            nd.clear();
            cn[i] = 0;
            dfs(i);
            int cur = 0;
            for(auto j : nd) {
                cur += abs(c[j] - cn[j]);
                cn[j] = -1;
            }

            nd.clear();
            cn[i] = 1;
            dfs(i);
            int cur2 = 0;
            for(auto j : nd) {
                cur2 += abs(c[j] - cn[j]);
            }
            ans += min(cur, cur2);
        }
    }
    cout << ans << endl;
}
int main() {
    // ios::sync_with_stdio(false);
    // cin.tie(nullptr);
    int t = 1;
    while(t--) solve();
    return 0;
}

File: contest/250304-sduwh-newstar/4.py
import sys
import threading
sys.setrecursionlimit(1000000)
# threading.stack_size(1024*1024*128)

n,m = map(int,input().split())
c = [0] + list(map(int,input().split()))

a = []
for i in range(n+5):
    a.append([])

for _ in range(m):
    x, y = map(int,input().split())
    if x == y:
        exit()
    a[x].append(y)
    a[y].append(x)

nd = []
cn = [-1] * (n+5)
result = True
def dfs(u: int) -> bool:
    # print(u)
    global nd
    nd.append(u)
    for v in a[u]:
        if cn[v] == -1:
            cn[v] = 1 - cn[u]
            dfs(v)
        elif cn[v] == cn[u]:
            # print(u, v)
            global result
            result = False

ans = 0
for i in range(1, n+1):
    if cn[i] == -1:
        nd.clear()
        cn[i] = 0
        dfs(i)
        if not result:
            print('Impossible')
            exit()
        cur = sum(abs(c[i] - cn[i]) for i in nd)
        for node in nd:
            cn[node] = -1

        nd = []
        cn[i] = 1
        dfs(i)
        if not result:
            print('Impossible')
            exit()
        cur = min(cur, sum(abs(c[i] - cn[i]) for i in nd))
        ans += cur

print(ans)
